doc/source/Reference.rst

```diff
@@ -162,16 +162,119 @@ Local Functions
+
+We highlight the steps down below.
+For the sake of argument let's imagine that the user is storing the chemistry data in a variable called ``my_chemistry``.
+
+1. First, the user should allocate ``my_chemistry`` and initialize the stored parameters with :c:func:`local_initialize_chemistry_parameters`.
+
+   .. code-block:: c++
+
+      chemistry_data * my_chemistry_data = new chemistry_data;
+      if (local_initialize_chemistry_parameters(my_chemistry_data) == 0) {
+         fprintf(stderr, "Error in local_initialize_chemistry_parameters.\n");
+      }
+
+2. Next, the user can configure the stored parameters.
+   They can do this by directly modifying the stored parameters (e.g. ``my_chemistry->use_grackle = 1``) or by using the :ref:`dynamic-api`.
+
+After the user has finished initializing ``my_chemistry``, and have configured an instance of :c:data:`code_units` (more detail provided :ref:`here <code-units>`), they can initialize an instance of :c:data:`chemistry_data_storage` with 
+:c:func:`local_initialize_chemistry_data`:
+
+.. code-block:: c++
+
+  chemistry_data_storage* my_rates = new chemistry_data_storage;
+  if (local_initialize_chemistry_data(my_chemistry, my_rates, &my_units) == 0) {
+    fprintf(stderr, "Error in local_initialize_chemistry_data.\n");
+    return 0;
+  }
+
+Configuration/Cleanup Functions
++++++++++++++++++++++++++++++++
+
+.. c:function:: int local_initialize_chemistry_parameters \
+                (chemistry_data *my_chemistry);
+
+   Initializes the parameters stored in the :c:type:`chemistry_data` data structure to their default values.
+   This should be called before run-time parameters are set.
+
+   This is the "local" counterpart to :c:func:`set_default_chemistry_parameters`.
+
+   :param chemistry_data \*my_chemistry: run-time parameters
+   :rtype: int
+   :returns: 1 (success) or 0 (failure)
+
+.. c:function:: int local_initialize_chemistry_data \
+                (chemistry_data *my_chemistry, \
+                chemistry_data_storage *my_rates, \
+                code_units *my_units);
+
+   Allocates storage for and initializes the values of all relevant chemistry and cooling rate data.
+   This data is stored within the provided :c:data:`chemistry_data_storage` struct.
+   This is the "local" counterpart to :c:func:`initialize_chemistry_data`.
+
+   This function should only be called after the user has finished configuring both ``my_chemistry`` and ``my_units``.
+   This function assumes that none of ``my_rates``'s members of pointer type hold valid memory addresses (i.e. where applicable, the function allocates fresh storage and makes no attempts to deallocate/reuse storage).
+
+   After calling this function, the user should avoid modifying any of the fields of ``my_chemistry`` or ``my_units``.
+   The **only** exception to this rule is that ``my_units``'s :c:var:`a_value` field may be mutated when using comoving units (i.e. ``my_units``'s :c:var:`comoving_coordinates` field is ``1``).
```

Comment:
I'm a little confused by what you mean here.

My impression was that:
- regardless of the value of ``my_units.comoving_coordinates``, the values of ``my_units.density_units`` and ``my_units.length_units`` affected the values of the rates as they are read into the ``chemistry_data_storage`` struct during the call to ``initialize_chemistry_data``
-  in other words they affect the units that are internally attached to the rates.
- Consequently, if you change the value of ``my_units.density_units`` or ``my_units.length_units`` between a call to ``initialize_chemistry_data`` and ``local_solve_chemistry`` the units used internally would be inconsistent.

To put this more concretely, my impression is that the following code snippet produce different outcomes based on the value of ``CONDITION``. Am I wrong?

```c++
const double true_length_units = 3.086e21;  // 1 comoving kpc

code_units my_units;
my_units.comoving_coordinates = 1;
my_units.density_units = 1.67e-24; // 1 m_H/comoving cc
if (CONDITION) {
  my_units.length_units = 2 * true_length_units;
} else {
  my_units.length_units = true_length_units;
}
/* < insert code that initializes:  my_units.time_units
                                    my_units.a_units
                                    my_units.a_value     > */

chemistry_data my_chem;
local_initialize_chemistry_parameters(&my_chem);
/* < insert code that initialize fields of my_chem ... > */
chemistry_data_storage my_rates;
local_initialize_chemistry_data(&my_chem, &my_rates, &my_units);

grackle_field_data my_fields;
/* < insert code that initializes attributes of my_fields > */ 


// when CONDITION==0, the following line has no effect
my_units.length_units = true_length_units;
local_solve_chemistry(&my_chem, &my_rates, &my_units, &my_fields, 0.5);
```

Reply:
Sorry, I forgot to follow up on this discussion. In a comoving coordinate unit system (at least as we define it here), we expect the following behavior:
```
x [code units, comoving frame] * X_UNITS = x [cgs units, proper frame]
```
In this situation, `length_units` ~ a and `density_units` ~ a^-3, so you then find that `cool_units` remains constant. It's been left the way it is in the code to preserve history, but it should probably be discussed somewhere. For here, you could maybe reword this to say that `my_chemistry` should not be modified and `my_units` should only be modified in a way that keeps the internal cooling units constant. I would be happy to write some additional text describing the internal cooling unit and how it should work with a comoving unit system.

Reply:
Ok, so I have pushed a new commit that replaced the following statement:

```rst
   After calling this function, the user should avoid modifying any of
   the fields of ``my_chemistry`` or ``my_units``.  The **only**
   exception to this rule is that ``my_units``'s :c:var:`a_value`
   field may be mutated when using comoving units (i.e. ``my_units``'s
   :c:var:`comoving_coordinates` field is ``1``).
```

with:
```rst
   After calling this function, the user should avoid modifying any of
   the fields of ``my_chemistry``.  The user should also avoid
   modifying ``my_units`` in a way that modifies the internal cooling
   units (e.g. it's fine to mutate ``my_units``'s :c:var:`a_value`
   field when its :c:var:`comoving_coordinates` field is ``1``).
```

I can get rid of the example in the parentheses if you would prefer. I only include it since I expect it's the most common modification that users may want to make to ``my_units``.

(On a more general note, I do think it would probably be useful if you write some additional text in a follow-up PR describing the internal cooling unit and its relationship with a comoving unit system)